### ISSUE TITLE
ODBC drivers incompatible with Informix 12.10

### ISSUE THREAD

**Issue by sn0wer** [ibmdb/node-ibm_db#678]
Hello,

I'm using ibm_db driver version 2.6.4 and Informix 12.10. I noticed that when using the driver, it did not correctly map the Informix datatype with Javascript. This occurs with DATETIME datatype that is returned as a string instead of an object.

This should be handled by the function GetColumns of odbc.cpp, which is using SQL_DESC_CONCISE_TYPE. Instead it should use SQL_DESC_TYPE (https://www.ibm.com/support/knowledgecenter/SSEPGG_11.5.0/com.ibm.db2.luw.apdv.cli.doc/doc/r0002218.html 
: "Sets SQL_DESC_TYPE and SQL_DESC_CONCISE_TYPE to the value of ValueType, except that if ValueType is one of the concise identifiers of a datetime, it sets SQL_DESC_TYPE to SQL_DATETIME, sets SQL_DESC_CONCISE_TYPE to the concise identifier, and sets SQL_DESC_DATETIME_INTERVAL_CODE to the corresponding datetime subcode.")

Thank you.

**Comment by sn0wer** [comment]
1/ Use Informix 12.10, NodeJS 11.15 and Loopback 3.0
2/ Create a table with a column that is a datatype DATETIME and insert some records
create table test1 (testdate datetime year to second);
insert into test1 (testdate) values ('2020-08-24 00:00:00');
3/ Create a model test1 in Loopback : 
`{
  "name": "test1",
  "plural": "test1",
  "base": "PersistedModel",
  "idInjection": true,
  "options": {
    "validateUpsert": true
  },
  "properties": {
    "testdate": {
      "type": "date",
      "required": true,
    }
  },
  "validations": [],
  "relations": {},
  "acls": [
    {
      "principalType": "ROLE",
      "accessType":"*",
      "principalId": "$unauthenticated",
      "permission": "DENY",
      "property": "*"
    }
  ],
  "methods": {}
}
`
4/ Query it (I'm using loopback-connector-informix and Loopback 3.0) : 
`
let query = "SELECT * from test1";
test1.dataSource.connector.query(query, (err, res) => {
console.log(typeof res[0].testdate);
});
`
5/ You'll see that it returns a string, and in your code it's supposed to return an object if it's DATETIME

**Comment by bimalkjha** [comment]
@sn0wer I got the issue. One work around would be let application convert the string to date object using new Date(String) method.
We can update ibm_db code to return date object instead of string, but worry is - it may break the existing application of users which is already processing the string value of date! Changing behavior will force them to change the application code.
We would like to know more from you. If work around works for you, it is good. Else, we need to change current behavior in some major release with documentation. Thanks.

**Comment by sn0wer** [comment]
@bimaljha I understand your point of view, you have to be careful not to break application of current users.
We have to apply manually the conversion, but I find this solution improper, especially since your code (https://github.com/ibmdb/node-ibm_db/blob/master/src/odbc.cpp#L490) is supposed to handle this case. Is it working on db2 but not on Informix (see SQL_DESC_CONCISE_TYPE) ?

**Comment by bimalkjha** [comment]
@sn0wer I looked into this issue. We have two test cases in ibm_db to test this scenario. You can execute `ibm_db\test\test-date.js` and `test-all-data-types.js` to verify it. It the result of these test files differs between Db2 and Informix, please share here.

The SQL_DESC_CONSIZE_TYPE returned for Db2 is as below:
```
Column Type     SQL_DESC_CONSIZE_TYPE    SQL_DESC_TYPE
DATE            SQL_TYPE_DATE            SQL_DATETIME
TIME            SQL_TYPE_TIME            SQL_DATETIME
TIMESTAMP       SQL_TYPE_TIMESTAMP       SQL_DATETIME
```
So, we can distinguish between DATE, TIME and TIMESTAMP columns of Db2 using SQL_DESC_CONSIZE_TYPE only.

If we change the code of `ODBC::GetColumnValue()` to process the timestamp related code, the returned data from `ibm_db` changes from String to JS Date, but it losses fraction value of timestamp too. Also, it is difficult to represent DATE and TIME column in non-string format. If we run the `test-all-data-types.js`, we can see the inserted data for timestamp column in '2015-09-10 10:16:33.770139', but on select, ibm_db would return it as `2015-09-10T10:16:33.770Z`. So, inserted data will not match with selected data.
Hence, we'll leave the ibm_db code at present state to keep the current behavior. Thanks.

**Comment by bimalkjha** [comment]
This issue is fixed in ibmdb@2.8.0. Thanks.
